src/content/profile-data.js

```diff
@@ -136,34 +136,93 @@ export function defaultThreadOrder(threads: Thread[]) {
+  if (
+    implementation === 'combined' &&
+    chargeToCallerList.length === 0 &&
+    pruneSubtreeList.length === 0
+  ) {
+    return thread;
+  }
+
+  // Prepare the filtering functions.
+  const filterFuncByImplementation = _getFuncImplementationFilter(thread, implementation);
+  const filterOutByFunc = funcIndex =>
+    filterFuncByImplementation(funcIndex) ||
+    chargeToCallerList.includes(funcIndex);
+  const pruneByFunc = funcIndex => pruneSubtreeList.includes(funcIndex);
+
+  // Apply all the filters to the thread.
+  return _applyFuncFilters(thread, filterOutByFunc, pruneByFunc);
+}
+
+function _getFuncImplementationFilter(thread, implementation: string): IndexIntoFuncTable => boolean {
   const { funcTable, stringTable } = thread;
 
   switch (implementation) {
     case 'cpp':
-      return _filterThreadByFunc(thread, funcIndex => {
-        // Return quickly if this is a JS frame.
+      return funcIndex => {
+        // Discard quickly if this is a JS frame.
         if (funcTable.isJS[funcIndex]) {
-          return false;
+          return true;
         }
         // Regular C++ functions are associated with a resource that describes the
         // shared library that these C++ functions were loaded from. Jitcode is not
         // loaded from shared libraries but instead generated at runtime, so Jitcode
-        // frames are not associated with a shared library and thus have no resource
+        // frames are not associated with a shared library and thus have no resource.
+        // These stack frames should be discarded when looking at C++ only.
         const locationString = stringTable.getString(funcTable.name[funcIndex]);
         const isProbablyJitCode =
           funcTable.resource[funcIndex] === -1 && locationString.startsWith('0x');
-        return !isProbablyJitCode;
-      });
+        return isProbablyJitCode;
+      };
     case 'js':
-      return _filterThreadByFunc(thread, funcIndex => funcTable.isJS[funcIndex]);
+      // Discard if not JS.
+      return funcIndex => !funcTable.isJS[funcIndex];
   }
-  return thread;
+  return () => false;
 }
 
-function _filterThreadByFunc(
+/**
+ * Take two arbitrary functions. The first filters out individual stack frames based
+ * on the function. The second prunes the entire subtree based on the the function.
+ *
+ * filterOutByFunc:
+ *
+ *          0                    0
+ *        /  \      Func 1     / | \
+ *      1     2      ->       3  4  2
+ *     /\     /\                   / \
+ *    3 4    5  6                 5  6
+ *
+ * pruneByFunc:
+ *
+ *          0                    0
+ *        /  \      Func 1       |
+ *      1     2      ->          2
+ *     /\     /\                / \
+ *    3 4    5  6              5  6
+ *
+ */
+function _applyFuncFilters(
   thread: Thread,
-  filter: IndexIntoFuncTable => boolean
+  filterOutByFunc: IndexIntoFuncTable => boolean,
+  pruneByFunc: IndexIntoFuncTable => boolean
```

Comment:
I think these should be called filterOutFrameByFunc and filterOutSubtreeByFunc or something along those lines.